### PR TITLE
Fixes (one more round towards v2)

### PR DESCRIPTION
**Pull request by glima**
Please help validate those (lack of) breaks were bugs.

### CODE REVIEW

File: src/modules/flow/json/json.c [hunk 1]
```diff
@@ -992,6 +993,7 @@ json_path_add_new_element(struct sol_flow_node *node, struct json_element *base_
             SOL_INT_CHECK(r, == -ENOMEM, r);
             SOL_INT_CHECK_GOTO(r, < 0, error_parse);
 
+            break;
```
Comment (edersondisouza) [inline src/modules/flow/json/json.c]: Maybe you should try to ping original authors of these codes - maybe they can remember some peculiarity. That said, I guess that _not_ having breaks is correct for json patch: note that both handle 'YXZ_BLOB' then falls through 'XYZ' case: it appears to prepare data for next case.

Comment (glima) [inline src/modules/flow/json/json.c]: Mm, maybe. @otaviobp ?

Comment (otaviobp) [inline src/modules/flow/json/json.c]: The reason to not having that break is the reason @edersondisouza commented. It is not an error. it is intended. I remember having tests that explores that, and they are suppose to fail after your changes. Have you tested that?

**Comment by glima** [comment]
* Otavio Pontes <notifications@github.com> [2017-05-11 14:11:54 -0700]:

> otaviobp commented on this pull request.
> 
> 
> 
> > @@ -992,6 +993,7 @@ json_path_add_new_element(struct sol_flow_node *node, struct json_element *base_
>              SOL_INT_CHECK(r, == -ENOMEM, r);
>              SOL_INT_CHECK_GOTO(r, < 0, error_parse);
>  
> +            break;
> 
> The reason to not having that break is the reason @edersondisouza commented. It is not an error. it is intended. I remember having tests that explores that, and they are suppose to fail after your changes. Have you tested that?

Maybe I missed a test suite. OK, then, I'll rewrite it commenting the
explicit pass through, then. Thanks, guys.

> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/solettaproject/soletta/pull/2364#discussion_r116103512

-- 
Gustavo Lima Chaves
Intel - Open Source Technology Center

**Comment by glima** [comment]
The tests were passing, just checked. Anyway, breaks that should not be there removed.

**Comment by glima** [comment]
Merging it at the end of the day

File: src/lib/comms/sol-http-server-impl-microhttpd.c [hunk 2]
```diff
@@ -1034,6 +1034,7 @@ free_request(struct sol_http_request *request)
         switch (param->type) {
         case SOL_HTTP_PARAM_POST_DATA:
             free((char *)param->value.data.filename.data);
+            break;
```
Comment (edersondisouza) [inline src/lib/comms/sol-http-server-impl-microhttpd.c]: I guess that @ceolin is the one who can validate this - but I guess that current code is correct as well: code after `FOREACH` does basically the same for `request->param` - frees unconditionally `param.value.key_value.value.data` and `param.value.key_value.key.data`, and if `SOL_HTTP_PARAM_POST_DATA`, frees `param.value.data.filename.data`.

Comment (glima) [inline src/lib/comms/sol-http-server-impl-microhttpd.c]: After the loop I see no problem, just wondering if inside it we have a double-free scenario.

Comment (glima) [inline src/lib/comms/sol-http-server-impl-microhttpd.c]: @ceolin ping

File: src/lib/flow/sol-flow-node-options.c [hunk 3]
```diff
@@ -781,7 +781,7 @@ split_option(const char *input, const char **key, unsigned int *key_len, const c
 {
     const char *equal = strchr(input, '=');
 
-    if (!equal || equal == input || equal + 1 == '\0')
+    if (!equal || equal == input || *equal + 1 == '\0')
```
Comment (ceolin) [inline src/lib/flow/sol-flow-node-options.c]: shouldn't be: `*(equal + 1) == '\0'`

File: src/lib/comms/sol-http-server-impl-microhttpd.c (hunk 2 above)
Comment (ceolin) [inline src/lib/comms/sol-http-server-impl-microhttpd.c]: Don't need this break. It'll cause memory leak.
If you really want to change it you will need to do:

```
    SOL_HTTP_PARAMS_FOREACH_IDX (&request->params, param, idx) {                                               
        switch (param->type) {                                                                                 
        case SOL_HTTP_PARAM_POST_DATA:
            free((char*)param->value.data.value.data);                                                   
            free((char *)param->value.data.key.data);          
            free((char *)param->value.data.filename.data);                                                     
            break;                                                                                             
        case SOL_HTTP_PARAM_POST_FIELD:                                                                        
            free((char *)param->value.key_value.value.data);                                                   
            free((char *)param->value.key_value.key.data);                                                     
            break;                                                                                             
        default:                                                                                               
            break;                                                                                             
        }                                                                                                      
    }    
```

To my mind, just let it as it is.

**Comment by ceolin** [comment]
Please, remove the commit with the break from the serie and check the comment about the sol-flow-node-options.c

**Comment by edersondisouza** [comment]
On Thu, 2017-06-08 at 13:32 -0700, Flávio Vinicius Alvares Ceolin
wrote:
> @ceolin commented on this pull request.
> In src/lib/comms/sol-http-server-impl-microhttpd.c:
> > @@ -1034,6 +1034,7 @@ free_request(struct sol_http_request
> *request)
>          switch (param->type) {
>          case SOL_HTTP_PARAM_POST_DATA:
>              free((char *)param->value.data.filename.data);
> +            break;
> Don't need this break. It'll cause memory leak.
> If you really want to change it you will need to do:
>     SOL_HTTP_PARAMS_FOREACH_IDX (&request->params, param, idx)
> {                                               
>         switch (param->type)
> {                                                                    
>              
>         case SOL_HTTP_PARAM_POST_DATA:
>             free((char*)param-
> >value.data.value.data);                                             
>       
>             free((char *)param->value.data.key.data);          
>             free((char *)param-
> >value.data.filename.data);                                          
>            
>             break;                                                   
>                                           
>         case
> SOL_HTTP_PARAM_POST_FIELD:                                           
>                              
>             free((char *)param-
> >value.key_value.value.data);                                        
>            
>             free((char *)param-
> >value.key_value.key.data);                                          
>            
>             break;                                                   
>                                           
>         default:                                                     
>                                           
>             break;                                                   
>                                           
>         }                                                            
>                                           
>     }    
> To my mind, just let it as it is.

To prevent this confusion in future, better put a comment making sure
fallthrough is desired.

File: src/lib/flow/sol-flow-node-options.c (hunk 3 above)
Comment (glima) [inline src/lib/flow/sol-flow-node-options.c]: Damn, thanks for that.

**Comment by glima** [comment]
Yeah, I'll just commit the fall-through comment and ship it. Thanks, guys.

**Comment by glima** [comment]
Merged. We're failing make samples, damn. Will look more closely into that later.

File: src/lib/flow/sol-flow-node-options.c (hunk 3 above)
Comment (barbieri) [inline src/lib/flow/sol-flow-node-options.c]: `equal[1]` is less ugly than `*(equal + 1)`

**Pull request closed without merging**